File: frontend/src/api/cd2.ts
import client from './client'

export interface Machine {
  id: number
  ten_may: string
  ma_may?: string
  loai_may: string
  sort_order: number
  active: boolean
  phan_xuong_id?: number
}

export interface ProductionLog {
  id: number
  production_order_id: number
  phieu_in_id?: number
  machine_id: number
  event_type: string
  quantity_ok?: number
  quantity_loi?: number
  quantity_setup?: number
  ghi_chu?: string
  created_at: string
  worker?: string
}

export interface TrackPayload {
  production_order_id: number
  so_lsx?: string
  machine_id: number
  event_type: 'start' | 'stop' | 'resume' | 'complete' | 'error'
  quantity_ok?: number
  quantity_loi?: number
  quantity_setup?: number
  ghi_chu?: string
}

export interface MayIn {
  id: number
  ten_may: string
  active: boolean
  sort_order: number
  phan_xuong_id?: number
  capacity?: number | null
}

export interface MaySauIn {
  id: number
  ten_may: string
  active: boolean
  sort_order: number
  phan_xuong_id?: number
}

export interface MayScan {
  id: number
  ten_may: string
  active: boolean
  sort_order: number
  phan_xuong_id?: number
  don_gia: number | null
}

export interface PrinterUser {
  id: number
  token_user: string
  token_password?: string
  rfid_key?: string | null
  shift?: number | null
  active: boolean
  machine_id?: number | null
  machine_name?: string
}

export interface WorkerSession {
  printer_user_id: number
  worker_name: string
  shift?: number | null
  machine_id: number
  machine_name: string
  loai_may: string
}

export interface PhieuIn {
  id: number
  so_phieu: string
  so_lsx: string
  production_order_id?: number | null
  ten_hang?: string
  ten_khach_hang?: string
  ma_kh?: string
  quy_cach?: string
  loai?: string
  loai_in?: string
  pp_ghep?: string
  so_don?: string
  ngay_lenh?: string
  ngay_giao_hang?: string
  ghi_chu_printer?: string
  ghi_chu_prepare?: string
  ghi_chu?: string
  so_luong_phoi: number
  trang_thai: string
  may_in_id?: number | null
  ten_may?: string
  nguoi_in_id?: number
  ten_nguoi_in?: string
  ghi_chu_in?: string
  created_at: string
  updated_at: string
  kho_tt?: number
  dai_tt?: number
  ths?: string
  // Kết quả in
  so_luong_in_ok?: number
  ngay_in?: string
  ca?: string
  so_luong_loi?: number
  so_luong_setup?: number
  so_lan_setup?: number
  ghi_chu_ket_qua?: string
  // Kết quả sau in
  ngay_sau_in?: string
  ca_sau_in?: string
  so_luong_sau_in_ok?: number
  so_luong_sau_in_loi?: number
  ghi_chu_sau_in?: string
  // Thời gian thực (timestamps)
  gio_bat_dau_in?: string
  gio_hoan_thanh?: string
  gio_bat_dau_dinh_hinh?: string
}

export interface CompletePayload {
  ngay_in?: string
  ca?: string
  so_luong_in_ok?: number
  so_luong_loi?: number
  so_luong_setup?: number
  so_lan_setup?: number
  ghi_chu_ket_qua?: string
}

export interface SauInPayload {
  ngay_sau_in?: string
  ca_sau_in?: string
  so_luong_sau_in_ok?: number
  so_luong_sau_in_loi?: number
  ghi_chu_sau_in?: string
}

export interface KanbanData {
  columns: Record<string, PhieuIn[]>
  may_ins: MayIn[]
}

export interface SauInKanbanData {
  may_sau_ins: MaySauIn[]
  cho_gang_may: PhieuIn[]
  machines: Record<string, PhieuIn[]>
}

export interface ScanLog {
  id: number
  may_scan_id: number
  so_lsx: string
  ten_hang?: string
  dai?: number
  rong?: number
  cao?: number
  kho_tt?: number
  dien_tich?: number
  so_luong_tp: number
  don_gia?: number
  tien_luong?: number
  nguoi_sx?: string
  created_at: string
}

export interface ScanLookupResult {
  so_lsx: string
  ten_hang?: string
  kho_tt?: number
  dai_tt?: number
  dien_tich_don_vi?: number
  dai?: number
  rong?: number
  cao?: number
}

export interface ShiftCa {
  id: number
  name: string
  leader?: string | null
  active: boolean
  phan_xuong_id?: number
}

export interface ShiftConfigItem {
  id: number
  may_in_id: number
  shift_ca_id: number
  ngay: string
  gio_lam?: number | null
  gio_bat_dau?: string | null
  gio_ket_thuc?: string | null
  nghi_1?: number | null
  nghi_2?: number | null
  ten_may?: string
  ten_ca?: string
}

export const TRANG_THAI_COLORS: Record<string, string> = {
  cho_in: 'orange',
  ke_hoach: 'blue',
  dang_in: 'volcano',
  cho_dinh_hinh: 'purple',
  sau_in: 'cyan',
  dang_sau_in: 'geekblue',
  hoan_thanh: 'green',
}

export const TRANG_THAI_LABELS: Record<string, string> = {
  cho_in: 'Chờ in',
  ke_hoach: 'Kế hoạch',
  dang_in: 'Đang in',
  cho_dinh_hinh: 'Chờ định hình',
  sau_in: 'Sau in',
  dang_sau_in: 'Đang sau in',
  hoan_thanh: 'Hoàn thành',
}

export interface KhoRow {
  production_order_id: number
  so_lenh: string
  ten_hang: string | null
  ten_khach_hang: string | null
  ten_phap_nhan_sx: string | null
  order_ten_phan_xuong: string | null
  warehouse_id: number | null
  ten_phan_xuong: string | null
  phan_xuong_id: number | null
  cong_doan: string | null
  co_in: boolean
  chieu_kho: number | null
  chieu_cat: number | null
  tong_nhap: number
  tong_xuat: number
  tong_chuyen_phoi: number
  ton_kho: number
  ton_kho_tai_nguon: number  // phôi còn tại kho Hoàng Gia chưa chuyển đi (chỉ có nghĩa với CD2 xưởng)
  ton_kho_tai_cd2: number    // phôi đã đến kho CD2 chưa sản xuất
  don_gia_noi_bo: number | null  // giá nội bộ chuyển kho (đ/tấm) — từ LSX, dùng cho hạch toán quản trị
  phieu_in_hien_tai: { so_phieu: string; trang_thai: string } | null
}

export const cd2Api = {
  // Dashboard & monitor
  getDashboard: (params?: any) => client.get('/cd2/dashboard', { params }),
  getMachinesStatus: (phanXuongId?: number) =>
    client.get<any[]>('/cd2/monitor/machines', { params: { phan_xuong_id: phanXuongId } }),

  // Phiếu in — list / kanban
  listPhieuIn: (params?: any) => client.get<PhieuIn[]>('/cd2/phieu-in', { params }),
  getPhieuIn: (id: number) => client.get<PhieuIn>(`/cd2/phieu-in/${id}`),
  getKanban: (params?: any) => client.get<KanbanData>('/cd2/kanban', { params }),
  getSauInKanban: (params?: any) => client.get<SauInKanbanData>('/cd2/sau-in-kanban', { params }),
  getHistoryPhieuIn: (params?: any) => client.get<PhieuIn[]>('/cd2/history/phieu-in', { params }),

  // Phiếu in — CRUD & actions
  createPhieuIn: (data: any) => client.post<PhieuIn>('/cd2/phieu-in', data),
  updatePhieuIn: (id: number, data: any) => client.put<PhieuIn>(`/cd2/phieu-in/${id}`, data),
  deletePhieuIn: (id: number) => client.delete(`/cd2/phieu-in/${id}`),
  movePhieuIn: (id: number, data: { trang_thai?: string; may_in_id?: number | null; sort_order?: number }) =>
    client.put(`/cd2/phieu-in/${id}/move`, data),
  startPrinting: (id: number) => client.post(`/cd2/phieu-in/${id}/start`),
  completePrinting: (id: number, data: CompletePayload) => client.post(`/cd2/phieu-in/${id}/complete`, data),
  startSauIn: (id: number, data: SauInPayload) => client.post(`/cd2/phieu-in/${id}/sau-in`, data),
  hoanThanh: (id: number) => client.post(`/cd2/phieu-in/${id}/hoan-thanh`),
  huyPhieu: (id: number) => client.post(`/cd2/phieu-in/${id}/huy`),
  assignSauIn: (id: number, maySauInId: number) =>
    client.post(`/cd2/phieu-in/${id}/assign-sau-in`, { may_sau_in_id: maySauInId }),
  batDauSauIn: (id: number) => client.post(`/cd2/phieu-in/${id}/bat-dau-sau-in`),
  traVeSauIn: (id: number) => client.post(`/cd2/phieu-in/${id}/tra-ve-sau-in`),
  createFromLenhSx: (orderId: number, target: 'in' | 'sau_in' | 'auto') =>
    client.post(`/cd2/phieu-in/tu-lenh-sx/${orderId}`, null, { params: { target } }),

  // Kho phôi
  getTonKhoLsx: () => client.get<KhoRow[]>('/cd2/ton-kho-lsx'),

  // Máy in
  listMayIn: (params?: any) => client.get<MayIn[]>('/cd2/may-in', { params }),
  createMayIn: (data: Partial<MayIn>) => client.post<MayIn>('/cd2/may-in', data),
  updateMayIn: (id: number, data: Partial<MayIn>) => client.put<MayIn>(`/cd2/may-in/${id}`, data),
  deleteMayIn: (id: number) => client.delete(`/cd2/may-in/${id}`),

  // Máy sau in
  listMaySauIn: (params?: any) => client.get<MaySauIn[]>('/cd2/may-sau-in', { params }),
  createMaySauIn: (data: Partial<MaySauIn>) => client.post<MaySauIn>('/cd2/may-sau-in', data),
  updateMaySauIn: (id: number, data: Partial<MaySauIn>) => client.put<MaySauIn>(`/cd2/may-sau-in/${id}`, data),
  deleteMaySauIn: (id: number) => client.delete(`/cd2/may-sau-in/${id}`),

  // Máy scan
  listMayScan: (params?: any) => client.get<MayScan[]>('/cd2/may-scan', { params }),
  createMayScan: (data: Partial<MayScan>) => client.post<MayScan>('/cd2/may-scan', data),
  updateMayScan: (id: number, data: Partial<MayScan>) => client.put<MayScan>(`/cd2/may-scan/${id}`, data),
  deleteMayScan: (id: number) => client.delete(`/cd2/may-scan/${id}`),

  // Scan log
  getScanHistory: (params?: any) => client.get<ScanLog[]>('/cd2/scan-logs/history-list', { params }),
  createScanLog: (data: {
    may_scan_id: number
    so_lsx: string
    ten_hang?: string
    dai?: number
    rong?: number
    cao?: number
    kho_tt?: number
    dien_tich?: number
    so_luong_tp: number
    don_gia?: number
    nguoi_sx?: string
  }) => client.post<ScanLog>('/cd2/scan-logs/submit', data),
  deleteScanLog: (id: number) => client.delete(`/cd2/scan-logs/delete/${id}`),
  // Tra cứu theo Số lệnh (Production Order) - Dành cho trang Scan Máy
  scanLookup: (code: string) => client.get<ScanLookupResult>(`/cd2/scan/lookup/${code}`),
  // Tra cứu theo Số phiếu (Phieu In) - Dành cho Mobile Tracking QR
  phieuLookup: (code: string) => client.get<any>(`/cd2/scan-lookup/${code}`),

  // Ca làm việc
  listShiftCa: (params?: any) => client.get<ShiftCa[]>('/cd2/shift/ca', { params }),
  createShiftCa: (data: { name: string; leader?: string }) => client.post<ShiftCa>('/cd2/shift/ca', data),
  updateShiftCa: (id: number, data: Partial<ShiftCa>) => client.put<ShiftCa>(`/cd2/shift/ca/${id}`, data),
  deleteShiftCa: (id: number) => client.delete(`/cd2/shift/ca/${id}`),

  // Lịch ca
  listShiftConfig: (params?: any) => client.get<ShiftConfigItem[]>('/cd2/shift/config', { params }),
  createShiftConfig: (data: {
    may_in_id: number
    shift_ca_id: number
    ngay: string
    gio_lam?: number
    gio_bat_dau?: string
    gio_ket_thuc?: string
    nghi_1?: number
    nghi_2?: number
  }) => client.post<ShiftConfigItem>('/cd2/shift/config', data),
  deleteShiftConfig: (id: number) => client.delete(`/cd2/shift/config/${id}`),

  // Machine login (no JWT required)
  machineLogin: (data: { token_user?: string; token_password?: string; rfid_key?: string }) =>
    client.post<WorkerSession>('/cd2/machine-login', data),

  // Người in
  listPrinterUser: (params?: any) => client.get<PrinterUser[]>('/cd2/config/printer-user', { params }),
  createPrinterUser: (data: Partial<PrinterUser> & { token_user: string; token_password: string }) =>
    client.post<PrinterUser>('/cd2/config/printer-user', data),
  updatePrinterUser: (id: number, data: Partial<PrinterUser>) =>
    client.put<PrinterUser>(`/cd2/config/printer-user/${id}`, data),
  deletePrinterUser: (id: number) => client.delete(`/cd2/config/printer-user/${id}`),

  // Máy sản xuất (Machine — loai_may)
  listMachines: (params?: any) => client.get<Machine[]>('/cd2/machines', { params }),
  createMachine: (data: Partial<Machine>) => client.post<Machine>('/cd2/machines', data),
  updateMachine: (id: number, data: Partial<Machine>) => client.put<Machine>(`/cd2/machines/${id}`, data),
  getMachineLogs: (machineId: number) => client.get<any[]>(`/cd2/machines/${machineId}/logs`),
  trackProduction: (data: TrackPayload) => client.post<{ ok: boolean; log_id: number }>('/cd2/track', data),
  getOrderProgress: (orderId: number) => client.get<ProductionLog[]>(`/cd2/progress/${orderId}`),
}

File: backend/app/models/cd2.py
from datetime import date, datetime
from decimal import Decimal
from sqlalchemy import Boolean, Date, DateTime, ForeignKey, Integer, Numeric, String, Text
from sqlalchemy.orm import Mapped, mapped_column, relationship
from app.database import Base


class Machine(Base):
    """Bảng máy móc dùng chung cho toàn bộ nhà máy."""
    __tablename__ = "machines"

    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    ten_may: Mapped[str] = mapped_column(String(100), nullable=False)
    ma_may: Mapped[str | None] = mapped_column(String(50), unique=True)
    # in | be | dan | ghim | can_mang | boi | phoi | khac
    loai_may: Mapped[str] = mapped_column(String(50), default="khac")
    sort_order: Mapped[int] = mapped_column(Integer, default=0)
    active: Mapped[bool] = mapped_column(Boolean, default=True)
    phan_xuong_id: Mapped[int | None] = mapped_column(Integer, ForeignKey("phan_xuong.id"), nullable=True)

    phan_xuong_obj = relationship("PhanXuong", foreign_keys=[phan_xuong_id])
    logs: Mapped[list["ProductionLog"]] = relationship("ProductionLog", back_populates="machine_obj")


class ProductionLog(Base):
    """Nhật ký chi tiết các sự kiện sản xuất tại máy (Start, Stop, Complete)."""
    __tablename__ = "production_logs"

    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    production_order_id: Mapped[int] = mapped_column(Integer, ForeignKey("production_orders.id"), nullable=False)
    phieu_in_id: Mapped[int | None] = mapped_column(Integer, ForeignKey("phieu_in.id"), nullable=True)
    machine_id: Mapped[int] = mapped_column(Integer, ForeignKey("machines.id"), nullable=False)

    # start | stop | resume | complete
    event_type: Mapped[str] = mapped_column(String(20), nullable=False)

    quantity_ok: Mapped[Decimal | None] = mapped_column(Numeric(12, 3))
    quantity_loi: Mapped[Decimal | None] = mapped_column(Numeric(12, 3))
    quantity_setup: Mapped[Decimal | None] = mapped_column(Numeric(12, 3))

    ghi_chu: Mapped[str | None] = mapped_column(Text)
    created_by: Mapped[int | None] = mapped_column(Integer, ForeignKey("users.id"))
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=datetime.utcnow)

    machine_obj: Mapped["Machine"] = relationship("Machine", back_populates="logs")
    production_order = relationship("ProductionOrder")
    phieu_in = relationship("PhieuIn")
    creator = relationship("User")


class MayScan(Base):
    __tablename__ = "may_scan"

    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    ten_may: Mapped[str] = mapped_column(String(50), nullable=False)
    sort_order: Mapped[int] = mapped_column(Integer, default=0)
    active: Mapped[bool] = mapped_column(Boolean, default=True)
    don_gia: Mapped[Decimal | None] = mapped_column(Numeric(12, 2))
    phan_xuong_id: Mapped[int | None] = mapped_column(Integer, ForeignKey("phan_xuong.id"), nullable=True)

    scan_logs: Mapped[list["ScanLog"]] = relationship("ScanLog", back_populates="may_scan_obj")
    phan_xuong_obj = relationship("PhanXuong", foreign_keys=[phan_xuong_id])


class ScanLog(Base):
    __tablename__ = "scan_log"

    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    may_scan_id: Mapped[int] = mapped_column(Integer, ForeignKey("may_scan.id"), nullable=False)
    so_lsx: Mapped[str] = mapped_column(String(50), nullable=False)
    ten_hang: Mapped[str | None] = mapped_column(String(255))
    dai: Mapped[Decimal | None] = mapped_column(Numeric(10, 2))
    rong: Mapped[Decimal | None] = mapped_column(Numeric(10, 2))
    cao: Mapped[Decimal | None] = mapped_column(Numeric(10, 2))
    kho_tt: Mapped[Decimal | None] = mapped_column(Numeric(10, 2))
    dien_tich: Mapped[Decimal | None] = mapped_column(Numeric(14, 4))   # tổng m² (đã nhân SL)
    so_luong_tp: Mapped[Decimal] = mapped_column(Numeric(12, 3), nullable=False)
    don_gia: Mapped[Decimal | None] = mapped_column(Numeric(12, 2))
    tien_luong: Mapped[Decimal | None] = mapped_column(Numeric(14, 2))  # = dien_tich * don_gia
    nguoi_sx: Mapped[str | None] = mapped_column(String(100))
    ghi_chu: Mapped[str | None] = mapped_column(Text)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=datetime.utcnow)
    created_by: Mapped[int | None] = mapped_column(Integer, ForeignKey("users.id"))

    may_scan_obj: Mapped["MayScan"] = relationship("MayScan", back_populates="scan_logs")
    creator = relationship("User")  # type: ignore[assignment]


class MaySauIn(Base):
    __tablename__ = "may_sau_in"

    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    ten_may: Mapped[str] = mapped_column(String(50), nullable=False)
    sort_order: Mapped[int] = mapped_column(Integer, default=0)
    active: Mapped[bool] = mapped_column(Boolean, default=True)
    phan_xuong_id: Mapped[int | None] = mapped_column(Integer, ForeignKey("phan_xuong.id"), nullable=True)

    phieu_ins: Mapped[list["PhieuIn"]] = relationship("PhieuIn", back_populates="may_sau_in_obj")
    phan_xuong_obj = relationship("PhanXuong", foreign_keys=[phan_xuong_id])


class MayIn(Base):
    __tablename__ = "may_in"

    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    ten_may: Mapped[str] = mapped_column(String(50), nullable=False)
    sort_order: Mapped[int] = mapped_column(Integer, default=0)
    active: Mapped[bool] = mapped_column(Boolean, default=True)
    capacity: Mapped[Decimal | None] = mapped_column(Numeric(10, 2))
    phan_xuong_id: Mapped[int | None] = mapped_column(Integer, ForeignKey("phan_xuong.id"), nullable=True)

    phieu_ins: Mapped[list["PhieuIn"]] = relationship("PhieuIn", back_populates="may_in_obj")
    phan_xuong_obj = relationship("PhanXuong", foreign_keys=[phan_xuong_id])


class PhieuIn(Base):
    __tablename__ = "phieu_in"

    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    so_phieu: Mapped[str] = mapped_column(String(30), unique=True, nullable=False)
    production_order_id: Mapped[int | None] = mapped_column(Integer, ForeignKey("production_orders.id"))
    may_in_id: Mapped[int | None] = mapped_column(Integer, ForeignKey("may_in.id"))
    may_sau_in_id: Mapped[int | None] = mapped_column(Integer, ForeignKey("may_sau_in.id"))

    # cho_in | ke_hoach | dang_in | cho_dinh_hinh | sau_in | hoan_thanh | huy
    trang_thai: Mapped[str] = mapped_column(String(20), default="cho_in")
    sort_order: Mapped[int] = mapped_column(Integer, default=0)

    ten_hang: Mapped[str | None] = mapped_column(String(255))
    ma_kh: Mapped[str | None] = mapped_column(String(50))
    ten_khach_hang: Mapped[str | None] = mapped_column(String(255))
    quy_cach: Mapped[str | None] = mapped_column(String(100))
    so_luong_phoi: Mapped[Decimal | None] = mapped_column(Numeric(12, 3))
    ngay_lenh: Mapped[date | None] = mapped_column(Date)
    loai_in: Mapped[str | None] = mapped_column(String(50))
    loai: Mapped[str | None] = mapped_column(String(50))          # Thùng, Hộp, Khay...
    ths: Mapped[str | None] = mapped_column(String(20))           # loại sóng: B, C, C-B
    pp_ghep: Mapped[str | None] = mapped_column(String(50))       # Dán, Đóng Ghim
    ghi_chu_printer: Mapped[str | None] = mapped_column(Text)
    ghi_chu_prepare: Mapped[str | None] = mapped_column(Text)
    so_don: Mapped[str | None] = mapped_column(String(50))
    ngay_giao_hang: Mapped[date | None] = mapped_column(Date)
    ghi_chu: Mapped[str | None] = mapped_column(Text)

    # Kết quả in
    ngay_in: Mapped[date | None] = mapped_column(Date)
    ca: Mapped[str | None] = mapped_column(String(20))
    so_luong_in_ok: Mapped[Decimal | None] = mapped_column(Numeric(12, 3))
    so_luong_loi: Mapped[Decimal | None] = mapped_column(Numeric(12, 3))
    so_luong_setup: Mapped[Decimal | None] = mapped_column(Numeric(12, 3))
    so_lan_setup: Mapped[int | None] = mapped_column(Integer)
    ghi_chu_ket_qua: Mapped[str | None] = mapped_column(Text)

    # Kết quả sau in
    ngay_sau_in: Mapped[date | None] = mapped_column(Date)
    ca_sau_in: Mapped[str | None] = mapped_column(String(20))
    so_luong_sau_in_ok: Mapped[Decimal | None] = mapped_column(Numeric(12, 3))
    so_luong_sau_in_loi: Mapped[Decimal | None] = mapped_column(Numeric(12, 3))
    ghi_chu_sau_in: Mapped[str | None] = mapped_column(Text)

    gio_bat_dau_in: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    gio_hoan_thanh: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    gio_bat_dau_dinh_hinh: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    gio_hoan_thanh_dinh_hinh: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))

    phan_xuong_id: Mapped[int | None] = mapped_column(Integer, ForeignKey("phan_xuong.id"), nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=datetime.utcnow)
    created_by: Mapped[int | None] = mapped_column(Integer, ForeignKey("users.id"))

    may_in_obj: Mapped["MayIn | None"] = relationship("MayIn", back_populates="phieu_ins")
    may_sau_in_obj: Mapped["MaySauIn | None"] = relationship("MaySauIn", back_populates="phieu_ins")
    production_order = relationship("ProductionOrder")  # type: ignore[assignment]
    creator = relationship("User")  # type: ignore[assignment]
    phan_xuong_obj = relationship("PhanXuong", foreign_keys=[phan_xuong_id])


class ShiftCa(Base):
    __tablename__ = "shift_ca"

    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    name: Mapped[str] = mapped_column(String(50), nullable=False)
    leader: Mapped[str | None] = mapped_column(String(100))
    active: Mapped[bool] = mapped_column(Boolean, default=True)
    phan_xuong_id: Mapped[int | None] = mapped_column(Integer, ForeignKey("phan_xuong.id"), nullable=True)

    configs: Mapped[list["ShiftConfig"]] = relationship("ShiftConfig", back_populates="shift_ca_obj")
    phan_xuong_obj = relationship("PhanXuong", foreign_keys=[phan_xuong_id])


class ShiftConfig(Base):
    __tablename__ = "shift_config"

    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    may_in_id: Mapped[int] = mapped_column(Integer, ForeignKey("may_in.id"), nullable=False)
    shift_ca_id: Mapped[int] = mapped_column(Integer, ForeignKey("shift_ca.id"), nullable=False)
    ngay: Mapped[date] = mapped_column(Date, nullable=False)
    gio_lam: Mapped[Decimal | None] = mapped_column(Numeric(5, 2))
    gio_bat_dau: Mapped[str | None] = mapped_column(String(10))
    gio_ket_thuc: Mapped[str | None] = mapped_column(String(10))
    nghi_1: Mapped[int | None] = mapped_column(Integer)
    nghi_2: Mapped[int | None] = mapped_column(Integer)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=datetime.utcnow)

    may_in_obj: Mapped["MayIn"] = relationship("MayIn")
    shift_ca_obj: Mapped["ShiftCa"] = relationship("ShiftCa", back_populates="configs")


class PrinterUser(Base):
    __tablename__ = "printer_user"

    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    rfid_key: Mapped[str | None] = mapped_column(String(100))
    token_user: Mapped[str] = mapped_column(String(100), nullable=False)
    token_password: Mapped[str] = mapped_column(String(255), nullable=False)
    shift: Mapped[int | None] = mapped_column(Integer)
    active: Mapped[bool] = mapped_column(Boolean, default=True)
    machine_id: Mapped[int | None] = mapped_column(Integer, ForeignKey("machines.id"), nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=datetime.utcnow)

    machine: Mapped["Machine"] = relationship("Machine", foreign_keys=[machine_id])
